src/ape_compile/__init__.py

```diff
@@ -23,6 +26,56 @@ class Config(PluginConfig):
     Source exclusion globs across all file types.
     """
 
+    cache_folder: Optional[Path] = None
+    """
+    Path to contract dependency cache directory (e.g. `contracts/.cache`)
+    """
+
+    @property
+    def base_dir(self) -> Path:
+        """The base directory for compilation file path references"""
+
+        # These should be initialized by plugin config loading and pydantic validators before the
+        # time this prop is accessed.
+        assert self._config_manager is not None
+        assert self.cache_folder is not None
+
+        # If the dependency cache folder is configured, to be outside of the contracts dir, we want
+        # to use the projects folder to be the base dir for copmilation.
+        if self._config_manager.contracts_folder not in self.cache_folder.parents:
+            return self._config_manager.PROJECT_FOLDER
+
+        # Otherwise, we're defaulting to contracts folder for backwards compatibility. Chaning this
```

Comment:
question: Thoughts on for 0.8 using the project path as the base regardless, as a breaking change?

Reply:
I have no issues with making this change.  We may want to document the old way somehow so users can reconstruct their project to match the old ways (or just instruct them to use an old version of ape and its deps, if possible). It's hard to get a sense how important repeatable builds are.

I also think it's worth discussing the idea in #1912. The functionality in it could be used for some kinds of build stages to assemble the compilation directory structure. Then it wouldn't matter where users put things. And we wouldn't need to really put a dependency cache anywhere semi-permanently. It's possible we could introduce this in a way to not be breaking.